app/scripts/controllers/network/network.js

```diff
@@ -249,21 +249,35 @@ export default class NetworkController extends EventEmitter {
       this.emit(NETWORK_EVENTS.INFURA_IS_UNBLOCKED);
     }
 
-    ethQuery.sendAsync({ method: 'net_version' }, (err, networkVersion) => {
-      const currentNetwork = this.getNetworkState();
-      if (initialNetwork === currentNetwork) {
-        if (err) {
-          this.setNetworkState('loading');
-          // keep network details in sync with network state
-          this.clearNetworkDetails();
-          return;
-        }
-
-        this.setNetworkState(networkVersion);
-        // look up EIP-1559 support
-        this.getEIP1559Compatibility();
+    let networkVersion;
+    try {
+      networkVersion = await new Promise((resolve, reject) => {
+        ethQuery.sendAsync({ method: 'net_version' }, (error, result) => {
+          if (error) {
+            reject(error);
+          } else {
+            resolve(result);
+          }
+        });
+      });
+    } catch (error) {
+      if (initialNetwork !== this.getNetworkState()) {
+        return;
       }
-    });
+
+      this.setNetworkState('loading');
+      // keep network details in sync with network state
+      this.clearNetworkDetails();
+      return;
+    }
+
+    if (initialNetwork !== this.getNetworkState()) {
+      return;
+    }
+
+    this.setNetworkState(networkVersion);
+    // look up EIP-1559 support
+    await this.getEIP1559Compatibility();
```

Comment:
Hmm, this is a bit less straightforward than the previous code. What if you added this at the top of the file:

``` javascript
function getNetworkVersion(ethQuery) {
  return ethQuery.sendAsync({ method: 'net_version' }, (error, result) => {
    if (error) {
      reject(error);
    } else {
      resolve(result);
    }
  });
}
```

And then changed this code to:

``` suggestion
    let networkVersion;
    let error;
    try {
      networkVersion = await getNetworkVersion(ethQuery);
    } catch (e) {
      console.error(e);
      error = e;
    }

    if (initialNetwork === this.getNetworkState()) {
      if (error) {
        this.setNetworkState('loading');
        // keep network details in sync with network state
        this.clearNetworkDetails();
        return;
      }

      this.setNetworkState(networkVersion);
      // look up EIP-1559 support
      await this.getEIP1559Compatibility();
    }
```

Reply:
Thanks, great suggestion, I will do that

Reply:
Done in https://github.com/MetaMask/metamask-extension/pull/16881/commits/7a1ebe1ace7237f1fd3fd949fe6bd9768fed2de5

Reply:
We still have the repeated `if (initialNetwork !== this.getNetworkState())` here but I guess that's fine, it only happens twice.

Reply:
Er, actually, would you mind simplifying this a bit. I think it could be confusing for someone glancing at this due to the three `return`s. It is easy to think think that we are calling `if (initialNetwork !== this.getNetworkState())` when in fact we are not.

Reply:
Do you have any suggestions on how this could be simplified?

In either the success or error case, we want to bail immediately if a network switch has happened. I'm not sure of a simpler way to do that. I also don't follow what you mean by that last sentence. There are only two branches, and we make that check in both.

Reply:
Sure, I'll make a new comment so it's more visible.